src/modules/pulumi.rs

```diff
@@ -0,0 +1,295 @@
+        log::trace!("Looking for package file in {:?}", path);
+        let dir = std::fs::read_dir(path).ok()?;
+        let goal = dir.filter_map(Result::ok).find(|path| {
+            path.file_name() == OsStr::new("Pulumi.yaml")
+                || path.file_name() == OsStr::new("Pulumi.yml")
+        });
+        if let Some(goal) = goal {
+            return Some(goal.path());
+        }
+    }
+    log::trace!("Did not find a Pulumi package file");
+    None
+}
+
+/// We get the name of the current stack.
+///
+/// Pulumi has no CLI option that is fast enough to get this for us, but finding
+/// the location is simple. We get it ourselves.
+fn stack_name(project_file: &Path, context: &Context) -> Option<String> {
+    let mut file = File::open(&project_file).ok()?;
+
+    let mut contents = String::new();
+    file.read_to_string(&mut contents).ok()?;
+    let name =
+        YamlLoader::load_from_str(&contents)
+            .ok()
+            .map(|yaml| -> Option<Option<String>> {
+                log::trace!("Parsed {:?} into yaml", project_file);
+                let yaml = yaml.into_iter().next()?;
+                yaml.into_hash().map(|mut hash| -> Option<String> {
+                    hash.remove(&Yaml::String("name".to_string()))?
+                        .into_string()
+                })
+            })???;
+    log::trace!("Found project name: {:?}", name);
+
+    let workspace_file = get_pulumi_workspace(context, &name, project_file)
+        .map(File::open)?
+        .ok()?;
+    log::trace!("Trying to read workspace_file: {:?}", workspace_file);
+    let workspace: serde_json::Value = match serde_json::from_reader(workspace_file) {
+        Ok(k) => k,
+        Err(e) => {
+            log::debug!("Failed to parse workspace file: {}", e);
+            return None;
+        }
+    };
+    log::trace!("Read workspace_file: {:?}", workspace);
+    workspace
+        .as_object()?
+        .get("stack")?
+        .as_str()
+        .map(ToString::to_string)
+}
+
+/// Calculates the path of the workspace settings file for a given pulumi stack.
+fn get_pulumi_workspace(context: &Context, name: &str, project_file: &Path) -> Option<PathBuf> {
+    let project_file = if cfg!(test) {
+        // Because this depends on the absolute path of the file, it changes in
+        // each test run. We thus mock it.
+        "test".to_string()
+    } else {
+        let mut hasher = Sha1::new();
+        hasher.update(project_file.to_str()?.as_bytes());
+        encode_to_hex(&hasher.finalize().to_vec())
+    };
+    let unique_file_name = format!("{}-{}-workspace.json", name, project_file);
+    let mut path = pulumi_home_dir(context)?;
+    path.push("workspaces");
+    path.push(unique_file_name);
+    Some(path)
+}
+
+const HEXTABLE: &[char] = &[
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
+];
+
+/// Encode a u8 slice into a hexadecimal string.
+fn encode_to_hex(slice: &[u8]) -> String {
```

Comment:
Not necessarily a blocker (we can add another PR afterwards if it would slow this down too much), but this feels general enough to be in `utils.rs` instead of hanging out here.

Reply:
I'm happy to move it